### PR TITLE
feat: add textDocument/onTypeFormatting support

### PR DESCRIPTION
**Pull request by sfaqer**
## Summary

Реализует LSP-метод `textDocument/onTypeFormatting` поверх существующего токен-based `FormatProvider`.

Триггеры:
- `"\n"` (Enter) — переформатирует **предыдущую** строку: нормализация регистра ключевых слов, расстановка пробелов вокруг операторов. Edit range — `(L, 0)..(L+1, 0)`, не пересекает позицию курсора.
- `";"` — форматирует текущую строку **до позиции курсора включительно**.
- Неподдерживаемые триггеры / Enter на первой строке / пустые строки → пустой список правок (no-op).

Реализация переиспользует существующий `getTextEdits` / `getNewText`, поэтому «бесплатно» получает корректную обработку унарного минуса, ASSIGN-индента, fluent-цепочек, локалей (RU/EN) и канонизации ключевых слов.

## Changes

- `FormatProvider.getOnTypeFormatting(...)` — новая точка входа (~60 строк).
- `BSLTextDocumentService.onTypeFormatting(...)` — override LSP-метода по шаблону `formatting`/`rangeFormatting`.
- `BSLLanguageServer.initialize` — объявление `DocumentOnTypeFormattingOptions("\n", [";"])`.

## Test plan

- [x] `FormatProviderTest` — 6 новых юнит-тестов (все зелёные):
  - Enter после `если х=1 тогда` → `Если х = 1 Тогда\n`;
  - Enter внутри процедуры — отступ сохраняется, регистр нормализуется;
  - Enter на пустой строке → `[]`;
  - Enter на первой строке файла → `[]`;
  - `;` после `х=1;` → `х = 1;` (range до курсора);
  - неизвестный триггер → `[]`.
- [x] `./gradlew test --tests FormatProviderTest` — 14/14 passed.
- [x] `./gradlew compileJava compileTestJava` — успешно.

## Notes

Существующая реализация в `vsc-language-1c-bsl` (`DocumentFormattingEditProvider.provideOnTypeFormattingEdits`) — регексп-фолбэк, регистрируется только при отключённом LS. Здесь — токен-based реализация в самом LS, поэтому она активна всегда и переиспользует уже отлаженный движок форматирования.

## Summary by CodeRabbit

* **New Features**
  * Added on-type automatic formatting: pressing Enter reformats the previous line; typing semicolon reformats the current line up to the cursor. The language server now advertises and supports on-type formatting requests with newline and semicolon triggers.
* **Tests**
  * Added tests covering on-type formatting behaviors (Enter, semicolon, unknown triggers) and edge cases to ensure correct edits and no-op conditions.
* **Documentation**
  * Updated docs to list on-type formatting and show supported trigger characters (\n and ;).

[Review Change Stack](https://app.coderabbit.ai/change-stack/1c-syntax/bsl-language-server/pull/3908?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: src/main/java/com/github/_1c_syntax/bsl/languageserver/providers/FormatProvider.java [hunk 1]
```diff
@@ -118,6 +120,72 @@ public List<TextEdit> getFormatting(DocumentFormattingParams params, DocumentCon
...
+          return false;
+        }
+        // Конец токена должен укладываться в диапазон до позиции курсора, иначе токены,
+        // выходящие за курсор (например многострочные литералы), дублируют свой хвост в replace.
+        long endColumn = (long) token.getCharPositionInLine() + token.getText().length();
+        return endColumn <= cutoffCharacter;
+      })
+      .toList();
+
+    boolean hasContent = tokens.stream()
```
Comment (nixel2007) [inline src/main/java/com/github/_1c_syntax/bsl/languageserver/providers/FormatProvider.java]: documentContext.getContent().isEmpty()?

File: src/main/java/com/github/_1c_syntax/bsl/languageserver/providers/FormatProvider.java [hunk 2]
```diff
@@ -118,6 +120,72 @@ public List<TextEdit> getFormatting(DocumentFormattingParams params, DocumentCon
...
+      targetLineLsp = position.getLine();
+      editRange = Ranges.create(targetLineLsp, 0, position.getLine(), position.getCharacter());
+      cutoffCharacter = position.getCharacter();
+    } else {
+      return Collections.emptyList();
+    }
+
+    int antlrLine = targetLineLsp + 1;
+
+    List<Token> tokens = documentContext.getTokens().stream()
```
Comment (nixel2007) [inline src/main/java/com/github/_1c_syntax/bsl/languageserver/providers/FormatProvider.java]: Дороговато... А как rangeFormatting делает?

File: src/main/java/com/github/_1c_syntax/bsl/languageserver/providers/FormatProvider.java (hunk 1 above)
Comment (sfaqer) [reply src/main/java/com/github/_1c_syntax/bsl/languageserver/providers/FormatProvider.java]: Проверка избыточна — внутри `getNewText` есть тот же short-circuit (`filteredTokens(tokens).isEmpty()` → `""`), а `getTextEdits` тогда возвращает `Collections.emptyList()`. Снял её в `df71bb90d`.

`documentContext.getContent().isEmpty()` напрямую тут не подойдёт: он проверяет пустоту всего документа, а нам важно — есть ли значимые токены *на целевой строке* (для Enter — на предыдущей строке, чтобы пустой Enter подряд не порождал no-op-edit).

File: src/main/java/com/github/_1c_syntax/bsl/languageserver/providers/FormatProvider.java (hunk 2 above)
Comment (sfaqer) [reply src/main/java/com/github/_1c_syntax/bsl/languageserver/providers/FormatProvider.java]: Согласен. `rangeFormatting` делает ровно так же (`documentContext.getTokens().stream().filter(...)`), но он вызывается по запросу пользователя — а `onTypeFormatting` срабатывает на каждый Enter / `;`, и обходить весь поток токенов на каждом нажатии заметно дороже.

В `df71bb90d` заменил стрим на binary search по `getTokens()` + локальный обход только хвоста нужной строки: токены ANTLR упорядочены по `line`, так что находим начало целевой строки за O(log N), читаем до первого токена со следующей линии. Hot path стал O(log N + K) вместо O(N), где K — число токенов на строке.

`rangeFormatting` оставил как есть — там кейс реже и диапазон по линиям шире (бинарка дала бы меньший выигрыш).

**Comment by nixel2007** [comment]
Спасибо!
